File: packages/cds-android/src/main/java/com/coinbase/cds/theme/Sizes.kt
package com.coinbase.cds.theme

import androidx.compose.runtime.Immutable
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Icon size scale. Rung names are the CDS scale keys.
 */
@Immutable
public class CdsIconSize internal constructor(
    public val xs: Dp,
    public val s: Dp,
    public val m: Dp,
    public val l: Dp,
) {
    /** Resolves a [CdsIconSizeToken]: `CdsTheme.iconSize[CdsIconSizeToken.M]`. */
    public operator fun get(token: CdsIconSizeToken): Dp = when (token) {
        CdsIconSizeToken.Xs -> xs
        CdsIconSizeToken.S -> s
        CdsIconSizeToken.M -> m
        CdsIconSizeToken.L -> l
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CdsIconSize) return false
        return CdsIconSizeToken.entries.all { this[it] == other[it] }
    }

    override fun hashCode(): Int {
        var result = 0
        for (token in CdsIconSizeToken.entries) result = 31 * result + this[token].hashCode()
        return result
    }

    override fun toString(): String = "CdsIconSize(xs=$xs, s=$s, m=$m, l=$l)"

    public companion object {
        /** The `cds-default` icon size scale. */
        public val Default: CdsIconSize = CdsIconSize(
            xs = 12.dp,
            s = 16.dp,
            m = 24.dp,
            l = 32.dp,
        )
    }
}

/**
 * Every rung of the [CdsIconSize] scale as an enumerable value. Resolve one against a theme with
 * [CdsIconSize.get]; iterate the whole scale with `CdsIconSizeToken.entries`.
 *
 * Entries may be added in a minor release, so this enum is not safe to match exhaustively --
 * include an `else` branch in any `when` over it.
 */
public enum class CdsIconSizeToken {
    Xs, S, M, L,
    ;

    /** The canonical CDS key (`m`), for labels and for parsing serialized themes. */
    public val tokenName: String get() = name.lowercase()
}

/**
 * Avatar size scale. Rung names are the CDS scale keys.
 */
@Immutable
public class CdsAvatarSize internal constructor(
    public val s: Dp,
    public val m: Dp,
    public val l: Dp,
    public val xl: Dp,
    public val xxl: Dp,
    public val xxxl: Dp,
) {
    /** Resolves a [CdsAvatarSizeToken]: `CdsTheme.avatarSize[CdsAvatarSizeToken.Xl]`. */
    public operator fun get(token: CdsAvatarSizeToken): Dp = when (token) {
        CdsAvatarSizeToken.S -> s
        CdsAvatarSizeToken.M -> m
        CdsAvatarSizeToken.L -> l
        CdsAvatarSizeToken.Xl -> xl
        CdsAvatarSizeToken.Xxl -> xxl
        CdsAvatarSizeToken.Xxxl -> xxxl
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CdsAvatarSize) return false
        return CdsAvatarSizeToken.entries.all { this[it] == other[it] }
    }

    override fun hashCode(): Int {
        var result = 0
        for (token in CdsAvatarSizeToken.entries) result = 31 * result + this[token].hashCode()
        return result
    }

    override fun toString(): String = "CdsAvatarSize(s=$s, m=$m, l=$l, xl=$xl, ...)"

    public companion object {
        /** The `cds-default` avatar size scale. */
        public val Default: CdsAvatarSize = CdsAvatarSize(
            s = 16.dp,
            m = 24.dp,
            l = 32.dp,
            xl = 40.dp,
            xxl = 48.dp,
            xxxl = 56.dp,
        )
    }
}

/**
 * Every rung of the [CdsAvatarSize] scale as an enumerable value. Resolve one against a theme with
 * [CdsAvatarSize.get]; iterate the whole scale with `CdsAvatarSizeToken.entries`.
 *
 * Entries may be added in a minor release, so this enum is not safe to match exhaustively --
 * include an `else` branch in any `when` over it.
 */
public enum class CdsAvatarSizeToken {
    S, M, L, Xl, Xxl, Xxxl,
    ;

    /** The canonical CDS key (`xl`), for labels and for parsing serialized themes. */
    public val tokenName: String get() = name.lowercase()
}

/**
 * Fixed measurements for CDS's form controls. Not a ramp -- each name measures one specific part of
 * one specific control.
 */
@Immutable
public class CdsControlSize internal constructor(
    public val checkboxSize: Dp,
    public val radioSize: Dp,
    public val switchWidth: Dp,
    public val switchHeight: Dp,
    public val switchThumbSize: Dp,
    public val tileSize: Dp,
) {
    /** Resolves a [CdsControlSizeToken]: `CdsTheme.controlSize[CdsControlSizeToken.CheckboxSize]`. */
    public operator fun get(token: CdsControlSizeToken): Dp = when (token) {
        CdsControlSizeToken.CheckboxSize -> checkboxSize
        CdsControlSizeToken.RadioSize -> radioSize
        CdsControlSizeToken.SwitchWidth -> switchWidth
        CdsControlSizeToken.SwitchHeight -> switchHeight
        CdsControlSizeToken.SwitchThumbSize -> switchThumbSize
        CdsControlSizeToken.TileSize -> tileSize
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CdsControlSize) return false
        return CdsControlSizeToken.entries.all { this[it] == other[it] }
    }

    override fun hashCode(): Int {
        var result = 0
        for (token in CdsControlSizeToken.entries) result = 31 * result + this[token].hashCode()
        return result
    }

    override fun toString(): String = "CdsControlSize(checkboxSize=$checkboxSize, ...)"

    public companion object {
        /** The `cds-default` control measurements. */
        public val Default: CdsControlSize = CdsControlSize(
            checkboxSize = 20.dp,
            radioSize = 20.dp,
            switchWidth = 52.dp,
            switchHeight = 32.dp,
            switchThumbSize = 30.dp,
            tileSize = 106.dp,
        )
    }
}

/**
 * Every measurement in [CdsControlSize] as an enumerable value. Resolve one against a theme with
 * [CdsControlSize.get]; iterate the whole set with `CdsControlSizeToken.entries`.
 *
 * Entries may be added in a minor release, so this enum is not safe to match exhaustively --
 * include an `else` branch in any `when` over it.
 */
public enum class CdsControlSizeToken {
    CheckboxSize, RadioSize, SwitchWidth, SwitchHeight, SwitchThumbSize, TileSize,
    ;

    /** The canonical CDS key (`checkboxSize`), for labels and for parsing serialized themes. */
    public val tokenName: String get() = name.replaceFirstChar { it.lowercase() }
}

File: packages/cds-ios/Sources/Theme/Sizes.swift
import CoreGraphics

/// Icon size scale, mirroring `theme.iconSize` in `defaultTheme.ts`.
public struct CDSIconSize: Sendable, Equatable {
    public var xs: CGFloat
    public var s: CGFloat
    public var m: CGFloat
    public var l: CGFloat

    public init(xs: CGFloat = 12, s: CGFloat = 16, m: CGFloat = 24, l: CGFloat = 32) {
        self.xs = xs; self.s = s; self.m = m; self.l = l
    }

    public func with(_ mutate: (inout CDSIconSize) -> Void) -> CDSIconSize {
        var copy = self
        mutate(&copy)
        return copy
    }

    /// Resolve an icon-size token: `theme.iconSize[.m]`. Pairs with ``CDSIconSizeToken``.
    public subscript(_ token: CDSIconSizeToken) -> CGFloat {
        switch token {
        case .xs: return xs
        case .s: return s
        case .m: return m
        case .l: return l
        }
    }

    public static let `default` = CDSIconSize()
}

/// Avatar size scale, mirroring `theme.avatarSize` in `defaultTheme.ts`.
public struct CDSAvatarSize: Sendable, Equatable {
    public var s: CGFloat
    public var m: CGFloat
    public var l: CGFloat
    public var xl: CGFloat
    public var xxl: CGFloat
    public var xxxl: CGFloat

    public init(
        s: CGFloat = 16, m: CGFloat = 24, l: CGFloat = 32,
        xl: CGFloat = 40, xxl: CGFloat = 48, xxxl: CGFloat = 56
    ) {
        self.s = s; self.m = m; self.l = l
        self.xl = xl; self.xxl = xxl; self.xxxl = xxxl
    }

    public func with(_ mutate: (inout CDSAvatarSize) -> Void) -> CDSAvatarSize {
        var copy = self
        mutate(&copy)
        return copy
    }

    /// Resolve an avatar-size token: `theme.avatarSize[.xl]`. Pairs with ``CDSAvatarSizeToken``.
    public subscript(_ token: CDSAvatarSizeToken) -> CGFloat {
        switch token {
        case .s: return s
        case .m: return m
        case .l: return l
        case .xl: return xl
        case .xxl: return xxl
        case .xxxl: return xxxl
        }
    }

    public static let `default` = CDSAvatarSize()
}

/// Control size scale, mirroring `theme.controlSize` in `defaultTheme.ts`.
public struct CDSControlSize: Sendable, Equatable {
    public var checkboxSize: CGFloat
    public var radioSize: CGFloat
    public var switchWidth: CGFloat
    public var switchHeight: CGFloat
    public var switchThumbSize: CGFloat
    public var tileSize: CGFloat

    public init(
        checkboxSize: CGFloat = 20, radioSize: CGFloat = 20,
        switchWidth: CGFloat = 52, switchHeight: CGFloat = 32,
        switchThumbSize: CGFloat = 30, tileSize: CGFloat = 106
    ) {
        self.checkboxSize = checkboxSize
        self.radioSize = radioSize
        self.switchWidth = switchWidth
        self.switchHeight = switchHeight
        self.switchThumbSize = switchThumbSize
        self.tileSize = tileSize
    }

    public func with(_ mutate: (inout CDSControlSize) -> Void) -> CDSControlSize {
        var copy = self
        mutate(&copy)
        return copy
    }

    /// Resolve a control measurement: `theme.controlSize[.checkboxSize]`. Pairs with ``CDSControlSizeToken``.
    public subscript(_ token: CDSControlSizeToken) -> CGFloat {
        switch token {
        case .checkboxSize: return checkboxSize
        case .radioSize: return radioSize
        case .switchWidth: return switchWidth
        case .switchHeight: return switchHeight
        case .switchThumbSize: return switchThumbSize
        case .tileSize: return tileSize
        }
    }

    public static let `default` = CDSControlSize()
}
